tensorflow_recommenders_addons/dynamic_embedding/python/ops/shadow_embedding_ops.py

```diff
@@ -0,0 +1,213 @@
+                         collections=collections,
+                         model_mode=model_mode,
+                         distribute_strategy=distribute_strategy,
+                         name=name)
+    exists = kwargs.get('exists', None)
+    if exists is None:
+      self.exists = resource_variable_ops.ResourceVariable(
+          (),
+          trainable=False,
+          collections=collections,
+          name=ids_name,
+          dtype=dtypes.bool,
+          shape=tensor_shape.TensorShape(None))
+    else:
+      self.exists = exists
+
+  def prefetch_values(self, update=False):
+    if self.params.bp_v2:
+      r, exists = self.params.lookup(self.ids, return_exists=True)
+      self.exists.assign(exists)
+      self.prefetch_values_op = self.transform(r)
+    else:
+      self.prefetch_values_op = self.transform(self.params.lookup(self.ids))
+    return self.prefetch_values_op
+
+  def _reset_ids(self, ids):
+    if not context.executing_eagerly():
+      collected_ops = []
+    with ops.control_dependencies([self.ids.assign(ids, use_locking=True)]):
+      prefetch_op = self.prefetch_values(update=True)
+      if not context.executing_eagerly():
+        collected_ops.append(prefetch_op)
+    for s in self._tracked_slots:
+      partial_reset_op = s._reset_ids(ids)
+      if not context.executing_eagerly():
+        collected_ops.append(partial_reset_op)
+    if not context.executing_eagerly():
+      return control_flow_ops.group(collected_ops)
+
+  def _gather_saveables_for_checkpoint(self):
+    self._reset_ids(array_ops.zeros((0,), dtype=self.params.key_dtype))
+    return {trackable.VARIABLE_VALUE_KEY: self}
+
+
+def embedding_lookup(
+    shadow,
+    ids,
+    partition_strategy=None,  # pylint: disable=unused-argument
+    name=None,
+    validate_indices=None,  # pylint: disable=unused-argument
+):
+  """
+  Shadow version of dynamic_embedding.embedding_lookup. It use existed shadow
+  variable to to embedding lookup, and store the result. No by-product will
+  be introduced in this call. So it can be decorated by `tf.function`.
+
+  Args:
+    shadow: A ShadowVariable object.
+    ids: A tensor with any shape as same dtype of params.key_dtype.
+    partition_strategy: No used, for API compatiblity with `nn.emedding_lookup`.
+    name: A name for the operation.
+    validate_indices: No used, just for compatible with nn.embedding_lookup .
+
+  Returns:
+    A tensor with shape [shape of ids] + [dim],
+      dim is equal to the value dim of params.
+      containing the values from the params tensor(s) for keys in ids.
+  """
+  ids = ops.convert_to_tensor(ids)
+
+  if shadow.ids.dtype != ids.dtype:
+    raise ValueError('{} ids is not matched with ShadowVariable with ids'
+                     ' {},'.format(ids.dtype, shadow.ids.dtype))
+
+  with ops.name_scope(name, "shadow_embedding_lookup"):
+    with ops.control_dependencies([shadow._reset_ids(ids)]):
+      return shadow.read_value(do_prefetch=True)
```

Comment:
Here will return a tensor with shape=<unknown> when the shape of ids is <None>

Reply:
solved with tf.reshape

Reply:
solved with tf.reshape

Reply:
solve with:
embeddings = tf.reshape(embeddings_flat, tf.concat([tf.shape(ids),tf.constant(self.embedding_size,shape=(1,))],0))